ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
-	}
+	mbfl_convert_filter *filter1, *filter2;
 
-	filter1 = NULL;
-	filter2 = NULL;
-	if (mbfl_convert_filter_get_vtbl(string->encoding, toenc) != NULL) {
-		filter1 = mbfl_convert_filter_new(string->encoding, toenc, mbfl_memory_device_output, 0, &device);
-	} else {
-		filter2 = mbfl_convert_filter_new(&mbfl_encoding_wchar, toenc, mbfl_memory_device_output, 0, &device);
-		if (filter2 != NULL) {
-			filter1 = mbfl_convert_filter_new(string->encoding, &mbfl_encoding_wchar, (int (*)(int, void*))filter2->filter_function, NULL, filter2);
-			if (filter1 == NULL) {
-				mbfl_convert_filter_delete(filter2);
-			}
-		}
-	}
+	create_encoding_conversion_filters(string->encoding, toenc, &device, &filter1, &filter2);
 	if (filter1 == NULL) {
+		/* Can't convert from `string->encoding` to `toenc` */
 		return NULL;
 	}
 
-	if (filter2 != NULL) {
-		filter2->illegal_mode = MBFL_OUTPUTFILTER_ILLEGAL_MODE_CHAR;
-		filter2->illegal_substchar = 0x3f;		/* '?' */
-	}
-
 	mbfl_memory_device_init(&device, string->len, (string->len >> 2) + 8);
-
-	/* feed data */
-	n = string->len;
-	p = string->val;
-	if (p != NULL) {
-		while (n > 0) {
-			if ((*filter1->filter_function)(*p++, filter1) < 0) {
-				break;
-			}
-			n--;
-		}
-	}
-
+	mbfl_filter_feed_data(string, filter1);
 	mbfl_convert_filter_flush(filter1);
 	mbfl_convert_filter_delete(filter1);
-	if (filter2 != NULL) {
-		mbfl_convert_filter_flush(filter2);
+	if (filter2) {
 		mbfl_convert_filter_delete(filter2);
 	}
 
+	result->encoding = toenc;
 	return mbfl_memory_device_result(&device, result);
 }
 
+/*
+ *  strlen
+ */
+size_t mbfl_strlen(const mbfl_string *string)
+{
+	return byte_offset_to_char_offset(string->encoding, string->val, string->len, string->len);
+}
 
 /*
- * identify encoding
+ *  strpos
  */
-const mbfl_encoding *
-mbfl_identify_encoding(mbfl_string *string, const mbfl_encoding **elist, int elistsz, int strict)
+size_t mbfl_strpos(mbfl_string *haystack, mbfl_string *needle, ssize_t offset, int reverse)
 {
-	int i, num, bad;
-	size_t n;
-	unsigned char *p;
-	mbfl_identify_filter *flist, *filter;
-	const mbfl_encoding *encoding;
-
-	/* flist is an array of mbfl_identify_filter instances */
-	flist = ecalloc(elistsz, sizeof(mbfl_identify_filter));
-
-	num = 0;
-	if (elist != NULL) {
-		for (i = 0; i < elistsz; i++) {
-			if (!mbfl_identify_filter_init2(&flist[num], elist[i])) {
-				num++;
-			}
-		}
-	}
+	ZEND_ASSERT(haystack->encoding == needle->encoding);
 
-	/* feed data */
-	n = string->len;
-	p = string->val;
-
-	if (p != NULL) {
-		bad = 0;
-		while (n > 0) {
-			for (i = 0; i < num; i++) {
-				filter = &flist[i];
-				if (!filter->flag) {
-					(*filter->filter_function)(*p, filter);
-					if (filter->flag) {
-						bad++;
-					}
-				}
-			}
-			if ((num - 1) <= bad && !strict) {
-				break;
-			}
-			p++;
-			n--;
+	unsigned char *offset_pointer = haystack->val;
+	if (offset != 0) {
+		long str_len = mbfl_strlen(haystack); /* `long` because next comparison must be signed */
```

Comment:
Indeed. It looks like this is the only use of `char_offset_to_byte_offset` where we care whether the requested offset was out of bounds. In all other cases, what is wanted is to clamp to the beginning or end of the string.

Sigh.

Reply:
I think I would like to analyze which encodings are subject to the AABB problem you mentioned below first. It may turn out that once all encodings affected by AABB are converted to an encoding which isn't, this problem will become easier to solve.

Reply:
Hmm. I think I understand better why `mbfl_strpos` was written to convert all other encodings to UTF-8. Almost *everything* is affected by AABB, except for UTF-8 and single-byte encodings like ASCII or ISO-8859-*.

However... if our search function jumped forward in increments of 2 or 4 bytes on fixed-width 2- or 4-byte encodings, then a lot more would not be affected and we could skip the conversion to UTF-8.

Hmm.

Reply:
> However... if our search function jumped forward in increments of 2 or 4 bytes on fixed-width 2- or 4-byte encodings, then a lot more would not be affected and we could skip the conversion to UTF-8.

That's true. Question there would be whether there are any 2/4 byte encodings that are common enough to bother optimizing in this way (this would require reimplementing zend_memnstr for different widths, which has a non-trivial slow-path).

Reply:
I think my conclusion is that either way, this is not the time to implement that optimization. For now:

- Special-case single-byte encodings, perhaps by delegating to the ordinary `strpos` function
- Convert everything else to UTF-8
- Bring back something like `mbfl_find_offset_utf8`

Reply:
OK, this is done.